tools/config/__init__.py

```diff
@@ -1282,14 +1288,40 @@ def load_resources(self, resources):
 
             # Add features while we find new ones
             features = set(self.get_features())
-            if features == prev_features:
+            requires = set(self.app_config_data.get("requires", []))
+            if features == prev_features and requires == prev_requires:
                 break
 
             resources.add_features(features)
 
             prev_features = features
+            prev_requires = requires
         self.validate_config()
-
+        missing_requirements = {}
+        for name, lib in self.lib_config_data.items():
+            for req in lib.get("requires", []):
+                if req not in self.lib_config_data:
+                    missing_requirements.setdefault(name, [])
+                    missing_requirements[name].append(req)
+        if missing_requirements:
+            message = "; ".join(
+                "library '{}' requires {} which is not present".format(
+                    name, ", ".join("'{}'".format(i) for i in missing)
+                )
+                for name, missing in missing_requirements.items()
+            )
+            raise ConfigException(message)
+        all_json_paths = [
+            cfg["__config_path"] for cfg in self.lib_config_data.values()
+        ]
+        included_json_files = [
+            ref for ref in resources.get_file_refs(FileType.JSON)
+            if (
+                not ref.name.endswith(self.__mbed_lib_config_name)
+                or abspath(ref.path) in all_json_paths
```

Comment:
Is this line saying: "Include all json files that are not `mbed_lib.json` files or are the `mbed_lib.json` files that have been specifically included by `requires`"?

If so, what is the case for the "Include all json files that are not `mbed_lib.json`"? Aren't we only interested in filtering on libraries?

Reply:
Resources object filters based on all `.json` files, as it has no understanding of what is or is not a library. This could allow us to ignore the not-used `mbed_app.json` file for example. I'm open for discussion on this design decision.

Reply:
I gotcha. I'd personally prefer to only include the `abspath(ref.path) in all_json_paths` condition in the first implementation and then make the condition more lax as the requirements dictate them.

Reply:
Let me see what that would look like real quick.

Reply:
Had to happen anyway see: https://github.com/ARMmbed/mbed-os/pull/9561/commits/db30e714f9eb4bc8f6c6de84438b396e6d1a2ba2